### PR TITLE
Adds `gradient_squared`  method` to `FiniteVolume`

### LINKED ISSUE #2979: Implement `gradient_squared` for `FiniteVolume`
### Description

At the moment `FiniteVolume` does not have a `gradient_squared` method, so it needs to be computed as the inner product of the gradient with itself.

### Motivation

Implementing it would allow some code simplifications in the heat sources (Ohmic & heat of mixing).

### Possible Implementation

Main issue is to find how we want to implement it (mathematically). If the discretised gradient of `u` is `L u`, then the gradient squared is just `u^T L^T L u`, where superscript `T` means transpose. Is this too naive?

### Additional context

_No Response_

### PR DESCRIPTION
**Pull request by medha-14**
# Description

Fixes #2979 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [ ] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by kratman** [comment]
@medha-14 This PR should also add tests and use the gradient squared function where appropriate

**Comment by medha-14** [comment]
I have written a test for the `gradient_squared` method . Could you please advise on where I should incorporate the use of the `gradient_squared` method ?

File: src/pybamm/spatial_methods/finite_volume.py [hunk 1]
```diff
@@ -128,6 +128,33 @@ def gradient_matrix(self, domain, domains):
...
+        -------
+        :class:`pybamm.Vector`
+            The gradient squared of the symbol.
+        """
+        domain = symbol.domain
+        gradient_matrix = self.gradient_matrix(domain, symbol.domains)
+
+        # Compute gradient squared: (∇u)^2 = u^T L^T L u
+        gradient_squared_matrix = gradient_matrix.T @ gradient_matrix
+        gradient_squared_result = gradient_squared_matrix @ discretised_symbol
```
Comment (brosaplanella) [inline src/pybamm/spatial_methods/finite_volume.py:154]: Shouldn't you have a discretised_symbol.T multiplying on the right?

File: src/pybamm/spatial_methods/finite_volume.py [hunk 2]
```diff
@@ -128,6 +128,33 @@ def gradient_matrix(self, domain, domains):
...
+
+        return gradient_squared_result
```
Comment (brosaplanella) [inline src/pybamm/spatial_methods/finite_volume.py:156]: Tests do not cover the new function, adding some tests to `test_finite_volume.py` (under the unit tests) should help. A good test would be to check that the gradient squared returns the same result than the inner product between two instances of the gradient.

**Review by brosaplanella — CHANGES_REQUESTED** [review]
Just some changes, mostly to improve coverages, but getting there!

File: src/pybamm/spatial_methods/finite_volume.py [hunk 3]
```diff
@@ -128,6 +128,35 @@ def gradient_matrix(self, domain, domains):
...
+        domain = symbol.domain
+        gradient_matrix = self.gradient_matrix(domain, symbol.domains)
+
+        # Compute gradient squared: (∇u)^2 = u^T (L^T L) u
+        gradient_squared_matrix = gradient_matrix.T @ gradient_matrix
+        gradient_squared_result = (
+            discretised_symbol.T @ gradient_squared_matrix @ discretised_symbol
+        )
+
+        return gradient_squared_result.item()
```
Comment (agriyakhetarpal) [inline src/pybamm/spatial_methods/finite_volume.py:158]: Here, `boundary_conditions` is mentioned in the function's signature, but it doesn't look like it's used in the logic of the function anywhere?

Comment (medha-14) [reply src/pybamm/spatial_methods/finite_volume.py]: Thank you for the feedback. I've updated the `gradient_squared` function to correctly handle boundary conditions, and I've adjusted the tests accordingly to verify these changes.

**Review by agriyakhetarpal — COMMENTED** [review]
Thanks, @medha-14!

**Comment by agriyakhetarpal** [comment]
I noticed that in the `gradient_matrix` method above the lines you touched, we could try building the gradient matrix directly in sparse format instead of converting to it later as a potential further improvement – by just pre-allocating all indices at once, and passing them to `scipy.sparse.block_diag` in favour of the current Kronecker product. Some part of the time there is spent in the array creation alongside the matrix construction, so maybe we can get a 2-4x improvement for larger matrices. Would you like to try that (not in this PR but in a separate one)? I'll be able to help, shall you need it.

**Comment by medha-14** [comment]
Any recommendations on how to improve the tests I have written for this one? or any other things I can change?

File: tests/unit/test_spatial_methods/test_scikit_finite_element.py [hunk 4]
```diff
@@ -157,6 +157,54 @@ def test_gradient(self):
...
+        gradient_squared = pybamm.grad_squared(var)
+        gradient_squared_disc = disc.process_symbol(gradient_squared)
+
+        inner_product_y = grad_disc_y.evaluate(None, 5 * y + 6 * z)
+        inner_product_z = grad_disc_z.evaluate(None, 5 * y + 6 * z)
+        inner_product = inner_product_y**2 + inner_product_z**2
+
+        grad_squared_result = gradient_squared_disc.evaluate(None, 5 * y + 6 * z)
+        np.testing.assert_array_almost_equal(grad_squared_result, inner_product)
+        np.testing.assert_array_less(0, grad_squared_result)
```
Comment (agriyakhetarpal) [inline tests/unit/test_spatial_methods/test_scikit_finite_element.py:187]: Could you please switch to `np.testing.assert_allclose()` here while we're at it? xref https://github.com/pybamm-team/pybamm/issues/4488

File: tests/unit/test_spatial_methods/test_scikit_finite_element.py [hunk 5]
```diff
@@ -157,6 +157,54 @@ def test_gradient(self):
...
+
+        gradient_squared_disc_dirichlet = disc.process_symbol(gradient_squared)
+        grad_squared_result_dirichlet = gradient_squared_disc_dirichlet.evaluate(
+            None, 5 * y + 6 * z
+        )
+
+        np.testing.assert_array_almost_equal(
+            grad_squared_result_dirichlet, inner_product
+        )
+        np.testing.assert_array_less(0, grad_squared_result_dirichlet)
```
Comment (agriyakhetarpal) [inline tests/unit/test_spatial_methods/test_scikit_finite_element.py:197]: Same as above

File: tests/unit/test_spatial_methods/test_scikit_finite_element.py [hunk 6]
```diff
@@ -157,6 +157,54 @@ def test_gradient(self):
...
+
+        gradient_squared_disc_neumann = disc.process_symbol(gradient_squared)
+        grad_squared_result_neumann = gradient_squared_disc_neumann.evaluate(
+            None, 5 * y + 6 * z
+        )
+
+        np.testing.assert_array_almost_equal(grad_squared_result_neumann, inner_product)
+
+        np.testing.assert_array_less(0, grad_squared_result_neumann)
```
Comment (agriyakhetarpal) [inline tests/unit/test_spatial_methods/test_scikit_finite_element.py:206]: Same as above

**Review by agriyakhetarpal — COMMENTED** [review]
Thanks, @medha-14! The code looks fine to me as long as coverage passes – I'm happy to approve it after that. I just have one additional comment.

**Comment by medha-14** [comment]
> Thanks, @medha-14! The code looks fine to me as long as coverage passes – I'm happy to approve it after that. I just have one additional comment.

I will prioritize fixing the coverage as soon as possible.

**Comment by kratman** [comment]
@medha-14 Is this still being worked on?

**Comment by medha-14** [comment]
Yes, I’m still on it . I’ve had a bit of a gap but will be picking it up again.
